### ISSUE TITLE
No performance improvement with Merbridge measurable

### ISSUE THREAD

**Issue by MerzMax** [merbridge/merbridge#291]
Hi, 

i am currently evaluating Istio together with Merbridge. Apparently I don’t see any improvements when using Merbridge compared to using plain Calico…

Except to the quite loud cloud noise, Merbridge seems to perform worse than Calico. If I understand it correctly, Merbridge should be minimum as fast as Calico and with higher loads even more performant. 

Can you explain this behavior? 

<img width="1014" alt="calico" src="https://user-images.githubusercontent.com/32777517/229447568-989c4fd1-7b18-4988-953a-9ee94be6cc37.png">

<img width="963" alt="merbridge" src="https://user-images.githubusercontent.com/32777517/229447597-a93b4727-259d-4cf1-b9bd-8fc04036a6a3.png">

Here are some Versions, that might help explaining the problem:

Istio: 1.16.2 (Installed with helm)
Calico: 3.24.5 (Installed with the Manifest yaml)
Merbridge: installed with `kubectl apply -f https://raw.githubusercontent.com/merbridge/merbridge/main/deploy/all-in-one.yaml`

**Comment by kebe7jun** [comment]
What tool did you use to test it?
Also, this m default profile does not enable cni mode by default, which may affect performance.

**Comment by MerzMax** [comment]
@kebe7jun Thanks for your reply! 

I use k6 as load test tool and httpbin as the serverside counter part. 

I also tried it with the CNI mode, wich did not show any improvements... The averages of the three tests i did were 26.87ms, avg=35.81ms, and avg=16.85ms.

**Comment by kebe7jun** [comment]
Can you provide more information? For example, your k6 configuration information, test case information, deployment architecture, etc. Help us to review and troubleshoot.

Theoretically this must be the wrong result and I need to figure out why ...... It would be best if the information you provide can help me reproduce the environment.

**Comment by kebe7jun** [comment]
After my testing, it is true that the max duration does increase in CNI mode, which may require further analysis to address. However, overall latency has been optimized to some extent.

I also tested the ambient mode, which further optimizes latency in this mode. Of course, the optimization effect of Merbridge for the same node is more pronounced at present.

```
without-merbridge-diff-node-test_results_summary.txt:921:     http_req_duration..............: avg=2.1ms   min=936.74µs med=1.97ms  max=31.77ms  p(90)=2.86ms  p(95)=3.15ms
2023-04-18_03-02-38_merbridge-ambient-diff-node-test_results_summary.txt:921:     http_req_duration..............: avg=2.05ms  min=890.7µs  med=1.9ms   max=15.43ms  p(90)=2.84ms  p(95)=3.18ms
2023-04-18_03-11-12_merbridge-main-cni-diff-node-test_results_summary.txt:921:     http_req_duration..............: avg=2.09ms  min=946.97µs med=1.96ms  max=34.72ms  p(90)=2.88ms  p(95)=3.15ms
2023-04-18_03-20-45_merbridge-main-cni-same-node-test_results_summary.txt:920:     http_req_duration..............: avg=1.9ms   min=842.19µs med=1.78ms  max=24.29ms  p(90)=2.58ms  p(95)=2.84ms
2023-04-18_03-30-53_without-merbridge-same-node-test_results_summary.txt:920:     http_req_duration..............: avg=1.98ms  min=856.5µs  med=1.88ms  max=17.62ms  p(90)=2.68ms  p(95)=2.92ms
2023-04-18_03-38-19_merbridge-ambient-same-node-test_results_summary.txt:921:     http_req_duration..............: avg=1.86ms  min=787.21µs med=1.74ms  max=19.97ms  p(90)=2.52ms  p(95)=2.78ms
```

ambient mode:
```
kubectl apply -f https://raw.githubusercontent.com/merbridge/merbridge/main/deploy/all-in-one.yaml
```

Note that, ambient profile enable debug mode by default, you should manually disable debug mode by modifing merbridge DaemonSet.

**Comment by MerzMax** [comment]
Hey @kebe7jun, 
first, thanks for investigating the issue! :) 

However, two questions have arisen for me:
1. In the [standard deployment file in line 72](https://github.com/merbridge/merbridge/blob/main/deploy/all-in-one.yaml#L72) the CNI mode is disabled. Therefore, if I understand it correctly, the tests my results are based on do not use this mode. As a result your explaination does not address the problems I determined.
2. What is the ambeint mode doing differently? Does this have anything to do with the Istio Ambient Mesh? In my setup I am using the standard Istio deployment with the sidecar proxies.

**Comment by kebe7jun** [comment]
The ambient mode is the addition of support for Istio ambient mode, which I have not yet merged into the main branch because it is an experimental feature.
In this mode, I did some refactoring of the underlying layer, which makes a big difference in handling the traffic forwarding logic, which can cause some performance impact.
Also, cni mode is no longer needed in ambient mode.

**Comment by kebe7jun** [comment]
> In the [standard deployment file in line 72](https://github.com/merbridge/merbridge/blob/main/deploy/all-in-one.yaml#L72) the CNI mode is disabled. Therefore, if I understand it correctly, the tests my results are based on do not use this mode. As a result your explaination does not address the problems I determined.

CNI mode optimizes connection performance.

**Comment by MerzMax** [comment]
I used the CNI mode in the past as well but the results did not show any improvement, compared to the "standard" Merbridge mode. 

Test1: 
> http_req_duration..............: avg=26.87ms  min=2.54ms  med=6.82ms   max=1.03s    p(90)=38.01ms  p(95)=111.13ms

Test2: 
> http_req_duration..............: avg=35.81ms  min=2.61ms  med=6.31ms  max=1.46s    p(90)=49.01ms  p(95)=167.29ms

Test3:
> http_req_duration..............: avg=16.85ms  min=2.55ms  med=5.83ms   max=1.25s    p(90)=26.12ms  p(95)=53.79ms

I will check, wheather I will see improvements with the experimental modeor not..

**Comment by MerzMax** [comment]
@kebe7jun I tested the ambient implementation and these are my results: 

Test1:
> http_req_duration..............: avg=17.65ms  min=2.33ms  med=4.71ms  max=800.27ms p(90)=24.61ms p(95)=61.99ms 

Test2: 
> http_req_duration..............: avg=16.4ms   min=2.35ms  med=4.7ms   max=733.69ms p(90)=29.11ms  p(95)=76.3ms  

Test 2:
> http_req_duration..............: avg=23.55ms  min=2.27ms  med=4.22ms  max=2s       p(90)=20.16ms  p(95)=63.01ms 

From my point of view this does not look like an improvement either :/

**Comment by kebe7jun** [comment]
what does Test* stand for？

**Comment by MerzMax** [comment]
@kebe7jun I just executed k6 3 times to get several test results.

**Comment by kebe7jun** [comment]
It looks like there is a large error in your results. I am not sure about the reliability of this test, as I conducted it in an environment with limited application. Perhaps other running applications could affect the performance?

Alternatively, could you provide a comparison between the results with and without Merbridge in an ambient mode? Could you please run the test three times?

**Comment by MerzMax** [comment]
@kebe7jun 

It could be that other applications affect it, but I have no idea which ones. On the other side, the results measured with "plain" calico and without Merbridge display high latency as well. If I understand eBPF and Merbridge correctly, the latency and CPU usage should be decrease when using Merbridge. As a result the request duration should be lower when using eBPF, which my tests with Merbridge can not verify. This is the case for both the standard and the ambient configuration. 

The test results from the standard Merbridge configuration can be found in the second chart of the initial issue. 

The results from using merbridge not in the am

**Comment by kebe7jun** [comment]
Yes, Merbridge does optimize network latency, but this effect is likely to be around 5-12% in a typical test scenario, which is likely to be submerged in error because the percentage difference is not significant.
The same test case, in my environment, the average latency is around 2ms, but in your environment it reaches 10ms+, which seems a bit abnormal, maybe related to the load of the nodes or other issues? (I deployed the environment exactly as you provided yaml).
I think the test results may be inaccurate if the test system is not providing stable service.

**Comment by MerzMax** [comment]
@kebe7jun You are right that the underlying infrastructure is not the best for load tests. Additionally I am surprised, that your setup's latency is less then 20% of my measured latency. 

Did you deploy the "cluster-flooder" to your setup and how did you setup your calico / other CNI beneath Merbridge?

**Comment by kebe7jun** [comment]
No cluster-flooder installed...

**Comment by MerzMax** [comment]
@kebe7jun This should make a hudge difference.. 
In my configurationt I run 250 Pods which are all registered as andpoints in 250 Services.

**Comment by kebe7jun** [comment]
My point is that the improvement in Merbridge is not significant, and when we really need to test performance metrics, there is no need to introduce other variables that will increase the error, which will prevent us from achieving the desired results of performance testing. Therefore, I did not deploy the workload.

**Comment by tanjunchen** [comment]
refer to https://github.com/merbridge/merbridge/issues/172

**Comment by tanjunchen** [comment]
@MerzMax  you can try https://github.com/istio/tools/tree/master/perf/benchmark  to test merbrdige.

**Comment by MerzMax** [comment]
@kebe7jun 
I want to test more than just plain eBPF versus iptables. In order to slow down the iptables check, a bunch of Services and Pods are needed to "spam" the rules. The cluster-flooder application accomplishes this. The cluster-flooder Pods are just Pods, including a container doing nothing (kubernetes/pause container). Therefore the applications should not increase the error. If they do so, they at least are increasing the error for the iptables and the eBPF test, which still allows comparability.
